### PR TITLE
Add param batchnorm layers

### PR DESCRIPTION
**Pull request by runjerry**
* Add 1d and 2d ParamBatchNorm layers

* Make configuration of param layers consistent with alf.nn.layers

* Update param layers and param networks accordingly.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -2073,45 +2190,380 @@ def forward(self, img, keep_group_dim=True):
...
             # reshape back: [B, n*C', H', W'] -> [B, n, C', H', W']
-            res = res.reshape(res.shape[0], self._groups, self._out_channels,
+            res = res.reshape(res.shape[0], self._n_groups, self._out_channels,
                               res.shape[2], res.shape[3])
 
         return res
 
 
+@alf.configurable
+class ParamBatchNorm(nn.Module):
```
Comment (emailweixu) [inline alf/layers.py]: now we have a separate file batch_norm.py. You can put this there.

File: alf/layers.py [hunk 2]
```diff
@@ -1757,25 +1759,42 @@ def __init__(self,
...
         else:
             self._bias_length = 0
             self._bias = None
+        if use_bn:
+            assert n_groups is not None, (
+                "n_groups has to be specified if use_bn")
+            self._bn = ParamBatchNorm1d(output_size, n_groups)
+            self._n_groups = n_groups
+        else:
+            n_groups = 1
```
Comment (emailweixu) [inline alf/layers.py:1800]: why is n_groups 1 if not use_bn?

File: alf/layers.py [hunk 3]
```diff
@@ -1941,18 +2012,28 @@ def __init__(self,
...
         else:
             self._bias_length = 0
             self._bias = None
+        if use_bn:
+            assert n_groups is not None, (
+                "n_groups has to be specified if use_bn")
+            self._bn = ParamBatchNorm2d(out_channels, n_groups)
+            self._n_groups = n_groups
+        else:
+            n_groups = 1
```
Comment (emailweixu) [inline alf/layers.py:2043]: similar to ParamFC, n_groups = 1?

File: alf/networks/param_networks.py [hunk 4]
```diff
@@ -183,20 +193,26 @@ def __init__(self,
                 tuple takes a format
                 ``(filters, kernel_size, strides, padding, pooling_kernel)``,
                 where ``padding`` and ``pooling_kernel`` are optional.
-            fc_layer_params (tuple[tuple]): a tuple of tuples where each tuple
-                takes a format ``(FC layer sizes. use_bias)``, where
-                ``use_bias`` is optional.
+            fc_layer_params (tuple[int]): a tuple of integers
```
Comment (emailweixu) [inline alf/networks/param_networks.py:196]: Perhaps a more generalizable way of doing this is to implement ParamSequential

File: alf/layers.py (hunk 1 above)
Comment (runjerry) [reply alf/layers.py]: Updated.

File: alf/layers.py (hunk 2 above)
Comment (runjerry) [reply alf/layers.py:1800]: If not ``use_bn``, then ``self._n_groups`` will be determined by input parameters when calling ``set_parameters``, i.e., we allow calling ParamFC with changing ``n_groups`` if not ``use_bn``. In this case, ``n_groups`` could be initialized as any integer. But we have to fix a ``self._n_groups`` if bn is used. 

Updated the docstring to explain it.

File: alf/layers.py (hunk 3 above)
Comment (runjerry) [reply alf/layers.py:2043]: yep, same reason as explained above.

File: alf/networks/param_networks.py (hunk 4 above)
Comment (runjerry) [reply alf/networks/param_networks.py:196]: Yeah, that would be ideal. I will think it over after the submission deadline.

**Comment by runjerry** [comment]
Merged into PR #1205
